Use of service principal for generating access tokens (#476)
microsoft/MLOS: mlos_bench/mlos_bench/services/remote/azure/README.md

# Azure Services

## Authentication

> This service allows us to generate tokens for authenticating with other Azure services.

Reviewer: ```suggestion
The `AzureAuthService` allows us to generate tokens for authenticating with other Azure services.

The goal is to be able to support long running experiments ideally without needing prompt the user to refresh their login session to avoid timeouts.
```

It supports authenticating as the following identities:

### User

As a user, the only requirement is that a user is logged in via the Azure CLI before `mlos_bench` is started i.e. `az login` or `az login --device-login`.

> This method is simpler to get started, but the drawback is that organization policies might limit the period in which we can generate access tokens under a user logged in via the CLI.

Reviewer: ```suggestion
This method is simpler to get started, but the drawback is that security policies might limit the period in which we can generate access tokens under a user logged in via the CLI.
```

> This would lead to early termination of an experiment once tokens cannot be generated anymore without another interactive login by the user.

Reviewer: ```suggestion
This would lead to early termination of an experiment once tokens cannot be generated anymore without another interactive login by the user.

To avoid this issue, we also support automated authentication as a Service Principle, described next.
```

### Service Principal (SP)

Similar to above, it also requires a user to be logged in first the Azure CLI to kick off the flow for the SP.

> However, this method addresses the drawback above to some extent as the login for a SP can be set to expire much later i.e. when the associated certificate expires in a year.

Reviewer: ```suggestion
However, this method addresses the drawback above to some extent as the since we can configure the system to automatically reauthenticate as the SP on demand and restrict its access to fewer resources to mitigate security concerns so that experiments can run as long as necessary.
```

The requirements of this method are:
- User logged in to Azure CLI

> - An existing SP with appropriate access to the target resource group i.e. `Contributor` role.

Reviewer: ```suggestion
- An existing SP with appropriate access to the target resource group (e.g., `Contributor` role).
  > See [scripts/setup-rg/README.md](./scripts/setup-rg/README.md) for details on setting up the SP.
```

- Key vault for storing certificate.
- A certificate associated with the SP, which is stored in the key vault.

> - User has access to read secrets / certificates from the key vault i.e. `Key Vault Administrator` role.

Reviewer: ```suggestion
- User has access to read the necessary secrets/certificates from the key vault (e.g., `Key Vault Administrator` role).
```

Though we should validate whether we can get away with some lesser privilege as well.

> The flow of authentication is as follows:

Reviewer: ```suggestion
Once the SP is setup, the flow of authentication within `mlos_bench` is as follows:
```

1. On initialization, retrieve the SP's certificate from the key vault as the current user.

> 2. Log in as the SP using the certificate

Reviewer: ```suggestion
2. Log in as the SP using the certificate.
```

> 3. During runtime when tokens are requested, we request them as the SP

Reviewer: ```suggestion
3. During runtime when tokens are requested, we request them as the SP, reauthenticating as necessary using the certificate which is stored only in memory.
```

The requirements of this method are:
- User logged in to Azure CLI
- An existing SP with appropriate access to the target resource group (e.g., `Contributor` role).
  > See [scripts/setup-rg/README.md](./scripts/setup-rg/README.md) for details on setting up the SP.
- Key vault for storing certificate.
- A certificate associated with the SP, which is stored in the key vault.
- User has access to read the necessary secrets/certificates from the key vault (e.g., `Key Vault Administrator` role).

Once the SP is setup, the flow of authentication within `mlos_bench` is as follows:
1. On initialization, retrieve the SP's certificate from the key vault as the current user and keep it in memory.
2. Log in as the SP using the certificate retrieved from the key vault.
3. During runtime when tokens are requested, we request them as the SP, reauthenticating as necessary using the certificate which is stored only in memory.

Although the user's login will expire earlier, it is only required on initialization.

> We retain the SP's credentials in memory while `mlos_bench` is running so that no interactive logins are required until the process ends.

Reviewer: These two sentences are a little redundant now given the text above.  Could probably remove them.

Author: Makes sense, removed them!
